### PR TITLE
Update add Review feature

### LINKED ISSUE #97: Update the Add Review Feature
To enable the user a quick start for a review and review from the course page,
The following PR's will be done:
1. Add a search field to the navbar, to enable a quick start.
2. Update the ```add_review``` page and form, so the course is selected in advance.
3. Improve both page's design with bootstrap.

### PR DESCRIPTION
**Pull request by oriAdler**
To enable the user a quick start for a review,
A search field was added to the navbar.
The user can search course by name. Afterward to choose the desired course
from a list of courses.

The following updates are to enable the new feature:
  1. Modify **urls.py** to add path to the new page
  2. Modify **models.py** to add new function ```get_courses_ordered_by_name```
  3. Add new template **add_review_search.html**
  4. Modify **app_layout.html** to add a search field to the navigation bar
  5. Add **add_review_search_tests.py**
  6. Modify **views.py** to add ```add_review_search``` function

fix: #97

**Notes:**
  1. If the user enters an empty string all courses returned
  2. At the moment, clicking on the course name makes a GET request
  to the ```add_review_page```, on the sequence PR
  clicking on the URL will make a GET request with the selected course name
  3. The page UI and navbar will be improved on a sequence PR

### CODE REVIEW

**Comment by oriAdler** [comment]
https://user-images.githubusercontent.com/57869309/117340598-bb4ca300-aea9-11eb-9b6b-6be58767de25.mp4

I've added a short video to demonstrate the new feature

File: homepage/templates/homepage/add_review_search.html [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+<body>
+{% extends 'homepage/app_layout.html' %}
+{% block content %}
+<div class="container-fluid text-left pt-5">
+  <div class="row">
+    <div class="col-sm-4">    
+        {% if courses %}
+        <h2>Click on Course to start a review</h2>
+        {% for course in courses %}
+            <a href="{% url 'add_review' %}">{{ course }}</a></br>
```
Comment (VeraTs) [inline homepage/templates/homepage/add_review_search.html:15]: Does this reroute to the page with the data already (filled in course name and professors possibilities) or is that up to further development?

**Review by VeraTs — COMMENTED** [review]
Looks Good! 
Just this little redirecting question

File: homepage/templates/homepage/add_review_search.html (hunk 1 above)
Comment (oriAdler) [reply homepage/templates/homepage/add_review_search.html:15]: At the moment, it reroutes to the general **add review** page without the data.
In the next PR, I'll update the **add review** page and modify that as you described.

File: homepage/tests/add_review_search_tests.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+@pytest.mark.django_db
+def test_renders_add_review_search_template(client):
+    response = client.get('/add_review_search/')
+
+    assert response.status_code == 200
+    assertTemplateUsed(response, 'homepage/add_review_search.html')
+
+
+@pytest.mark.django_db
+def test_get_course_by_name_with_client(client):
```
Comment (LeonRabinovich) [inline homepage/tests/add_review_search_tests.py:58]: Too many asserts IMO for a single function.
I would split this into 2 methods - 1 method for found courses and 1 method for not found courses.

File: homepage/tests/add_review_search_tests.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    response = client.get('/add_review_search/', {'course': 'Resonance'})
+    courses_not_found = [b'Grammatica in Arithmancy', b'Numerology', b'UnFogging the Future']
+
+    assert response.status_code == 200
+    assert b'Resonance' in response.content
+    assert all(course not in response.content for course in courses_not_found)
```
Comment (LeonRabinovich) [inline homepage/tests/add_review_search_tests.py:58]: Too many asserts IMO for a single function.
I would split this into 2 methods - 1 method for found courses and 1 method for not found courses.

Comment (oriAdler) [reply homepage/tests/add_review_search_tests.py:58]: @LeonRabinovich Please take a closer look. 
When the user searches for a course, the expected result is that single course.
Therefore we must check that the rest of the courses are absent in the response.
BTW you can see the function for not found courses below.

Comment (LeonRabinovich) [reply homepage/tests/add_review_search_tests.py:58]: @oriAdler 
Not sure what you mean, i meant for you to simply split the same method you already have into 2 methods.

(1)
**def test_get_course_by_name_with_client(client):**

    response = client.get('/add_review_search/', {'course': 'Resonance'})
    assert response.status_code == 200
    assert b'Resonance' in response.content

(2)
**def test_courses_not_found_with_client(client):**

    response = client.get('/add_review_search/', {'course': 'Resonance'})
    courses_not_found = [b'Grammatica in Arithmancy', b'Numerology', b'UnFogging the Future']
    assert all(course not in response.content for course in courses_not_found)

Comment (oriAdler) [reply homepage/tests/add_review_search_tests.py:58]: @LeonRabinovich I understand very well what you said in the first comment.
It makes no sense to split it into 2 tests as these checks are logically linked.
About the "Too many asserts", I invite you to take a look at the last lecture:
image
The code in the presentation has the same amount of asserts or even more, so it is completely fine.

**Review by VeraTs — APPROVED** [review]

File: homepage/tests/add_review_search_tests.py (hunk 3 above)
Comment (LeonRabinovich) [reply homepage/tests/add_review_search_tests.py:58]: @oriAdler 
IMO it does make sense to split it into 2 methods. the same can be said for a valid-invalid test which is split into 2 methods (1 for valid and 1 for invalid ).
in this case, 1 test for found course and 1 test for not found courses.

Regarding the number of asserts, i invite you to take a look at what Kobi said in my PR: #90 
image

Comment (oriAdler) [reply homepage/tests/add_review_search_tests.py:58]: @LeonRabinovich **The example you mentioned is different, as you were checking two aspects in the same test.**
As I said before, if you look again at the code, **there is already a separate test for not found courses**.
I explained in the first comment why these checks must be in the same test. 
**You've said you didn't understand the explanation, but you insist on the change**. I find that weird. 
Please read the first comment and tell me what you didn't understand so I can explain.

Comment (ellaml) [reply homepage/tests/add_review_search_tests.py:58]: @LeonRabinovich, I agree with @oriAdler IMO those asserts are linked, and test the same results from different aspects - that the course was found, and only that course. Therefore, they should be together in the same test.

File: homepage/tests/add_review_search_tests.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+
+@pytest.mark.django_db
+def test_wrong_course_name_with_client(client):
+    response = client.get('/add_review_search/', {'course': 'Introduction to Buddhism'})
+
+    assert response.status_code == 200
+    assert b'oops!' in response.content
```
Comment (ellaml) [inline homepage/tests/add_review_search_tests.py]: This test will easily fail if we decide to change the content of the answer. maybe we can send back a different response code or somehow check that no courses are shown.
IMO we can keep only the first assertion to assure that an empty response is not redirecting us to a 404 page or something.

File: homepage/tests/add_review_search_tests.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    assert all(isinstance(course, Course) for course in courses)
+
+
+@pytest.mark.parametrize("course_name_exist, expected_output", [
+    ("Resonance", ["Resonance in Runes and Signs"]),
+    ("Numer", ["Numerology"]),
+    ("in", ["Grammatica in Arithmancy", "Resonance in Runes and Signs", "UnFogging the Future"]),
+])
+@pytest.mark.django_db
+def test_name_exist_get_courses_ordered_by_name(course_name_exist, expected_output):
```
Comment (ellaml) [inline homepage/tests/add_review_search_tests.py]: exist -> exists (optional)

File: homepage/tests/add_review_search_tests.py (hunk 4 above)
Comment (oriAdler) [reply homepage/tests/add_review_search_tests.py]: I've tried to change the status code to **204 no content**, but the browser didn't render the response.
So as you've suggested, I've kept only the first assertion.

File: homepage/tests/add_review_search_tests.py (hunk 5 above)
Comment (oriAdler) [reply homepage/tests/add_review_search_tests.py]: Changed to exists

**Review by ellaml — APPROVED** [review]
LGTM

File: homepage/tests/add_review_search_tests.py (hunk 3 above)
Comment (LeonRabinovich) [reply homepage/tests/add_review_search_tests.py:58]: @ellaml @oriAdler 
Ok, i approved the review

**Review by LeonRabinovich — APPROVED** [review]

File: homepage/tests/add_review_search_tests.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+from django.db.models import QuerySet
+from pytest_django.asserts import assertTemplateUsed
+
+
+# --------Backend testing-------- #
+@pytest.mark.django_db
+def test_type_get_courses_ordered_by_name():
+    # because an empty string is a substring of all names, all courses returned
+    courses = Course.get_courses_ordered_by_name("")
+
```
Comment (kobihk) [inline homepage/tests/add_review_search_tests.py]: please remove the blank line.
I would use a blank line as a "separator" if the method is too long and I have too many cases to separate them but in cases of a small method I think we don't have to add it.

File: homepage/tests/add_review_search_tests.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+@pytest.mark.parametrize("course_name_exist, expected_output", [
+    ("Resonance", ["Resonance in Runes and Signs"]),
+    ("Numer", ["Numerology"]),
+    ("in", ["Grammatica in Arithmancy", "Resonance in Runes and Signs", "UnFogging the Future"]),
+])
+@pytest.mark.django_db
+def test_name_exists_get_courses_ordered_by_name(course_name_exist, expected_output):
+    courses = Course.get_courses_ordered_by_name(course_name_exist)
+
```
Comment (kobihk) [inline homepage/tests/add_review_search_tests.py]: same ^^

File: homepage/tests/add_review_search_tests.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+@pytest.mark.parametrize("wrong_course_name", [
+    ("Introduction to Poetry"),
+    ("Introduction to Buddhism"),
+])
+@pytest.mark.django_db
+def test_wrong_name_get_courses_ordered_by_name(wrong_course_name):
+    courses = Course.get_courses_ordered_by_name(wrong_course_name)
+
```
Comment (kobihk) [inline homepage/tests/add_review_search_tests.py]: same ^^

File: homepage/tests/add_review_search_tests.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    assert not courses
+
+
+@pytest.mark.django_db
+def test_lexicography_order_get_courses_ordered_by_name():
+    courses = list(Course.get_courses_ordered_by_name(""))
+
```
Comment (kobihk) [inline homepage/tests/add_review_search_tests.py]: same ^^

File: homepage/tests/add_review_search_tests.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    for index in range(len(courses)-1):
+        assert courses[index].name < courses[index+1].name
+
+
+# --------Front End testing-------- #
+@pytest.mark.django_db
+def test_renders_add_review_search_template(client):
+    response = client.get('/add_review_search/')
+
```
Comment (kobihk) [inline homepage/tests/add_review_search_tests.py]: same ^^

File: homepage/tests/add_review_search_tests.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    assert response.status_code == 200
+    assertTemplateUsed(response, 'homepage/add_review_search.html')
+
+
+@pytest.mark.django_db
+def test_get_course_by_name_with_client(client):
+    response = client.get('/add_review_search/', {'course': 'Resonance'})
+    courses_not_found = [b'Grammatica in Arithmancy', b'Numerology', b'UnFogging the Future']
+
```
Comment (kobihk) [inline homepage/tests/add_review_search_tests.py]: same ^^

File: homepage/tests/add_review_search_tests.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    assert response.status_code == 200
+    assert b'Resonance' in response.content
+    assert all(course not in response.content for course in courses_not_found)
+
+
+@pytest.mark.django_db
+def test_wrong_course_name_with_client(client):
+    response = client.get('/add_review_search/', {'course': 'Introduction to Buddhism'})
+
```
Comment (kobihk) [inline homepage/tests/add_review_search_tests.py]: same ^^

**Review by kobihk — CHANGES_REQUESTED** [review]

**Review by kobihk — APPROVED** [review]

**Pull request merged**